File: LC/[0727] Minimum Window Subsequence - finite state machine.cpp
class Solution {
    public:
        string minWindow(string s1, string s2) {
            //min substring of s1 s.t. s2 is a subsequence.
    
            int const m = s1.size();//take s1 measurement
            s1="#"+s1;//insert dummy
    
            //declare state machine
            int next[m+1][26];//index of ch in s1 after s1[ii]
    
            //initialize state machine
            for (int ch = 0; ch<26; ++ch) next[m][ch]=-1; //nothing after last char in s1
    
            for(int ii=m-1; ii>=0; --ii){//from (m-1) walking backward as m is initializd in above line
                for(int ch=0; ch<26; ++ch){ //for each lowercase letter
                    next[ii][ch] = next[ii+1][ch];
                }
                next[ii][s1[ii+1]-'a']=ii+1;
            }
    
            //identify starts
            vector<int> start;
            for(int ii=1; ii<=m; ++ii) if(s1[ii]==s2[0]) start.push_back(ii);
            
            string ret = "";
            for(int ii: start){//for each start
                int jj = ii-1; //set observation index, we actually already know ii is a valid starting point
                int flag=1;
     
                for(auto ch: s2){//for each char in s2, use "next" to identify start, end index
                    jj = next[jj][ch-'a'];
                    if(jj==-1) {
                        flag=0;
                        break;
                    }
                }
                if(flag){
                    int len = jj-ii+1;
                    if(ret=="" || len < ret.size()){
                        ret = s1.substr(ii, len);
                    }
                }
              
            }
            return ret;   
        }
    };

File: LC/[0727] Minimum Window Subsequence - finite state machine.java
class Solution {
    public String minWindow(String s1, String s2) {
        //take s1,s2 metrics
        int N1=s1.length();
        int N2=s2.length();

        //insert dummy
        s1="#"+s1;
        s2="#"+s2;

        //declare finite-state-machine
        var next = new int[N1+1][26];

        //initialize next
        //what's the know initial state -> no next @ end of s1
        for(int ch=0; ch<26;++ch) next[N1][ch]=-1;//-1 indicate no next
        //build all states from initial state
        for(int ii=N1-1; ii>=0; --ii){//ii=N1 is covered by above line, going backward
            for(int ch=0; ch<26;++ch)
                next[ii][ch]=next[ii+1][ch];//check out of bound on ii+1
            next[ii][s1.charAt(ii+1)-'a'] = ii+1;//s1[ii]'s next idx of s1[ii+1] is ii+1
        }
        
        //collect all starts
        var starts = new ArrayList<Integer>();
        for(int ii=1; ii<=N1; ++ii) if(s1.charAt(ii)==s2.charAt(1)) starts.add(ii);
        
        //for each start find the substring and maintain the shortest one
  
        String res="";
        int len = Integer.MAX_VALUE;
        for(var st: starts){
            Boolean flag = true;//assuming we could find the subsequence from this st
            var ed = st;//end index from s1
            for(int jj=2; jj<=N2; ++jj){//jj is index into s2
                ed=next[ed][s2.charAt(jj)-'a'];//find subsequence from s2
                //System.out.println("st: " + st + " ed: " + ed);
                if(ed==-1){//can we find all index in s1 that covers s2?
                    flag=false;
                    break;
                }
            }
            //System.out.println(" out of loop: ed: " + ed);
            if(flag && (ed-st+1) < len) {//if we have a qualified substring and shorter
                res=s1.substring(st, ed+1);
                len=ed-st+1;
            }
        }
        return res;        
    }
}
/*
Problem statements:
1. find min substring from s1 s.t. s2 is a subsequence

Analysis:
* two series -> DPIII: M*N
* Bruteforce: 
  for each start, two pointers/greedy to find shortest substring of S1 s.t. s2 is a subsequence.  S*m*n
* finite-state-machine:
   pre-build next[ii][ch]: the idx of ch in s1 after s[ii]
   s*(N)+26*N
*/
